src/main/java/com/google/devtools/build/lib/analysis/platform/ConstraintCollection.java

```diff
@@ -0,0 +1,134 @@
+import com.google.devtools.build.lib.skylarkbuildapi.platform.ConstraintCollectionApi;
+import com.google.devtools.build.lib.skylarkinterface.SkylarkPrinter;
+import com.google.devtools.build.lib.syntax.EvalException;
+import com.google.devtools.build.lib.syntax.EvalUtils;
+import com.google.devtools.build.lib.syntax.Printer;
+import com.google.devtools.build.lib.util.Fingerprint;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/** A collection of constraint values. */
+@Immutable
+@AutoCodec
+public class ConstraintCollection
+    implements ConstraintCollectionApi<ConstraintSettingInfo, ConstraintValueInfo> {
+  private final ImmutableMap<ConstraintSettingInfo, ConstraintValueInfo> constraints;
+
+  ConstraintCollection(ImmutableList<ConstraintValueInfo> constraints) {
+    this(
+        constraints.stream()
+            .collect(
+                ImmutableMap.toImmutableMap(ConstraintValueInfo::constraint, Function.identity())));
+  }
+
+  @AutoCodec.Instantiator
+  @VisibleForSerialization
+  ConstraintCollection(ImmutableMap<ConstraintSettingInfo, ConstraintValueInfo> constraints) {
+    this.constraints = constraints;
+  }
+
+  private ConstraintSettingInfo convertKey(Object key, Location loc) throws EvalException {
+    if (!(key instanceof ConstraintSettingInfo)) {
+      throw new EvalException(
+          loc,
+          String.format(
+              "Constraint names must be platform_common.ConstraintSettingInfo, got %s instead",
+              EvalUtils.getDataTypeName(key)));
+    }
+
+    return (ConstraintSettingInfo) key;
+  }
+
+  /**
+   * Returns the {@link ConstraintValueInfo} for the given {@link ConstraintSettingInfo}, or {@code
+   * null} if none exists.
+   */
+  @Nullable
+  @Override
+  public ConstraintValueInfo get(ConstraintSettingInfo constraint) {
+    return constraints.get(constraint);
+  }
+
+  @Override
+  public ImmutableSet<ConstraintSettingInfo> constraintSettings() {
+    return constraints.keySet();
+  }
+
+  @Override
+  public Object getIndex(Object key, Location loc) throws EvalException {
+    ConstraintSettingInfo constraint = convertKey(key, loc);
+    return get(constraint);
+  }
+
+  @Override
+  public boolean containsKey(Object key, Location loc) throws EvalException {
+    ConstraintSettingInfo constraint = convertKey(key, loc);
+    return constraints.containsKey(constraint);
+  }
+
+  @Override
+  public boolean equals(Object other) {
+    if (!(other instanceof ConstraintCollection)) {
+      return false;
+    }
+    ConstraintCollection pc = (ConstraintCollection) other;
+    return Objects.equal(this.constraints, pc.constraints);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(constraints);
+  }
+
+  @Override
+  public String toString() {
+    return Printer.str(this);
+  }
+
+  @Override
+  public void repr(SkylarkPrinter printer) {
+    printer.append("<");
+    printer.append(
+        constraints.values().stream()
```

Comment:
nit: why not print the keys, too?

Reply:
ConstraintValueInfo.toString() already includes the setting label as well as the value label, so it's not needed.